File: kotlin/src/main/kotlin/com/pantas/debug/controlplane/DebugAuth.kt
package com.pantas.debug.controlplane

import java.security.MessageDigest
import java.time.Instant

/** Route auth class used by the debug-plane auth gate. */
enum class DebugAuthRouteClass {
    HELLO_BOOTSTRAP,
    AUTH_BOOTSTRAP,
    SENSITIVE,
}

/** Request facts needed by a pure JVM [DebugAuthManager]. */
data class DebugAuthRequest(
    val method: String,
    val segments: List<String>,
    val routeClass: DebugAuthRouteClass = DebugAuth.classifyRoute(method, segments),
    val bearerToken: String? = null,
    val body: Map<String, Any?> = emptyMap(),
    val request: Any? = null,
)

/** Token metadata stored by the App side. Plaintext tokens must not be stored here. */
data class DebugAuthTokenRecord(
    val tokenId: String,
    val tokenHash: String,
    val createdAt: Instant,
    val expiresAt: Instant,
    val revokedAt: Instant? = null,
    val clientLabel: String? = null,
)

/** Authorization decision shared by future route and SSE gates. */
sealed class DebugAuthDecision {
    data object Authorized : DebugAuthDecision()

    data class Denied(
        val statusCode: Int,
        val code: String,
        val message: String,
    ) : DebugAuthDecision()
}

/** Pluggable App-side authorizer. */
interface DebugAuthManager {
    suspend fun authorize(request: DebugAuthRequest): DebugAuthDecision

    suspend fun helloAuthState(token: String?): Map<String, Any?>
}

/** Compatibility manager used when the host has not enabled debug auth. */
object NoOpDebugAuthManager : DebugAuthManager {
    override suspend fun authorize(request: DebugAuthRequest): DebugAuthDecision =
        DebugAuthDecision.Authorized

    override suspend fun helloAuthState(token: String?): Map<String, Any?> =
        mapOf(
            "authRequired" to false,
            "authStatus" to "disabled",
        )
}

/** Pure helpers for auth route classification, Bearer parsing, and token validation. */
object DebugAuth {
    fun classifyRoute(method: String, segments: List<String>): DebugAuthRouteClass {
        val normalizedMethod = method.uppercase()
        return when {
            normalizedMethod == "GET" && segments == listOf("hello") ->
                DebugAuthRouteClass.HELLO_BOOTSTRAP

            normalizedMethod == "POST" &&
                segments.size == 2 &&
                segments[0] == "auth" &&
                segments[1] in AUTH_BOOTSTRAP_ACTIONS ->
                DebugAuthRouteClass.AUTH_BOOTSTRAP

            else -> DebugAuthRouteClass.SENSITIVE
        }
    }

    fun bearerToken(headers: Map<String, String>): String? {
        val value = headers.entries.firstOrNull { (key, _) ->
            key.equals("Authorization", ignoreCase = true)
        }?.value
        return bearerToken(value)
    }

    fun bearerToken(headerValue: String?): String? {
        if (headerValue == null) return null
        if (!headerValue.startsWith("Bearer ")) return null
        val token = headerValue.removePrefix("Bearer ")
        if (token.isBlank()) return null
        if (token.any { it.isWhitespace() }) return null
        return token
    }

    fun tokenHash(token: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
            .digest(token.toByteArray(Charsets.UTF_8))
        return digest.joinToString(separator = "") { byte -> "%02x".format(byte) }
    }

    fun validateToken(
        token: String?,
        record: DebugAuthTokenRecord?,
        now: Instant,
    ): DebugAuthDecision {
        if (token.isNullOrBlank()) return authorizationRequired()
        if (record == null) return invalidToken()
        if (record.revokedAt != null) return tokenRevoked()
        if (!now.isBefore(record.expiresAt)) return tokenExpired()
        if (tokenHash(token) != record.tokenHash) return invalidToken()
        return DebugAuthDecision.Authorized
    }

    fun authorizationRequired(): DebugAuthDecision.Denied =
        DebugAuthDecision.Denied(401, "authorization_required", "Debug authorization is required.")

    fun invalidToken(): DebugAuthDecision.Denied =
        DebugAuthDecision.Denied(401, "invalid_token", "Debug authorization token is invalid.")

    fun tokenExpired(): DebugAuthDecision.Denied =
        DebugAuthDecision.Denied(401, "token_expired", "Debug authorization token expired.")

    fun tokenRevoked(): DebugAuthDecision.Denied =
        DebugAuthDecision.Denied(401, "token_revoked", "Debug authorization token was revoked.")

    private val AUTH_BOOTSTRAP_ACTIONS = setOf("request", "status", "claim")
}

File: dart/lib/src/debug_auth.dart
enum AuthRouteClass {
  helloBootstrap,
  authBootstrap,
  sensitive,
}

class AuthRequest {
  const AuthRequest({
    required this.method,
    required this.segments,
    required this.routeClass,
    this.bearerToken,
    this.body = const <String, Object?>{},
    this.request,
  });

  final String method;
  final List<String> segments;
  final AuthRouteClass routeClass;
  final String? bearerToken;
  final Map<String, Object?> body;
  final Object? request;
}

sealed class AuthDecision {
  const AuthDecision();
}

class AuthAuthorized extends AuthDecision {
  const AuthAuthorized();
}

class AuthDenied extends AuthDecision {
  const AuthDenied({
    required this.statusCode,
    required this.code,
    required this.message,
  });

  final int statusCode;
  final String code;
  final String message;
}

sealed class AuthRouteResult {
  const AuthRouteResult();
}

class AuthRouteOk extends AuthRouteResult {
  const AuthRouteOk(this.body, {this.statusCode = 200});

  final Map<String, Object?> body;
  final int statusCode;
}

class AuthRouteDenied extends AuthRouteResult {
  const AuthRouteDenied({
    required this.statusCode,
    required this.code,
    required this.message,
  });

  final int statusCode;
  final String code;
  final String message;
}

abstract interface class DebugAuthManager {
  Future<AuthDecision> authorize(AuthRequest request);

  Future<Map<String, Object?>> helloAuthState(String? token);

  Future<AuthRouteResult> requestAuthorization(Map<String, Object?> body);

  Future<AuthRouteResult> authorizationStatus(Map<String, Object?> body);

  Future<AuthRouteResult> claimAuthorization(Map<String, Object?> body);
}

class NoOpDebugAuthManager implements DebugAuthManager {
  const NoOpDebugAuthManager();

  @override
  Future<AuthDecision> authorize(AuthRequest request) async =>
      const AuthAuthorized();

  @override
  Future<Map<String, Object?>> helloAuthState(String? token) async =>
      const <String, Object?>{
        'authRequired': false,
        'authStatus': 'disabled',
      };

  @override
  Future<AuthRouteResult> requestAuthorization(
    Map<String, Object?> body,
  ) async =>
      AuthRouteDenied(
        statusCode: 401,
        code: 'authorization_required',
        message: 'Debug authorization is required.',
      );

  @override
  Future<AuthRouteResult> authorizationStatus(
    Map<String, Object?> body,
  ) async =>
      AuthRouteDenied(
        statusCode: 401,
        code: 'invalid_token',
        message: 'Debug authorization token is invalid.',
      );

  @override
  Future<AuthRouteResult> claimAuthorization(
    Map<String, Object?> body,
  ) async =>
      AuthRouteDenied(
        statusCode: 401,
        code: 'invalid_token',
        message: 'Debug authorization token is invalid.',
      );
}

final class DebugAuth {
  const DebugAuth._();

  static AuthRouteClass classifyRoute(String method, List<String> segments) {
    final normalizedMethod = method.toUpperCase();
    if (normalizedMethod == 'GET' &&
        segments.length == 1 &&
        segments.first == 'hello') {
      return AuthRouteClass.helloBootstrap;
    }
    if (normalizedMethod == 'POST' &&
        segments.length == 2 &&
        segments.first == 'auth' &&
        _authBootstrapActions.contains(segments[1])) {
      return AuthRouteClass.authBootstrap;
    }
    return AuthRouteClass.sensitive;
  }

  static String? bearerToken(Map<String, String> headers) {
    for (final entry in headers.entries) {
      if (entry.key.toLowerCase() == 'authorization') {
        return bearerTokenValue(entry.value);
      }
    }
    return null;
  }

  static String? bearerTokenValue(String? headerValue) {
    if (headerValue == null || !headerValue.startsWith('Bearer ')) {
      return null;
    }
    final token = headerValue.substring('Bearer '.length);
    if (token.trim().isEmpty) return null;
    if (token.contains(RegExp(r'\s'))) return null;
    return token;
  }

  static AuthDenied authorizationRequired() => const AuthDenied(
        statusCode: 401,
        code: 'authorization_required',
        message: 'Debug authorization is required.',
      );

  static AuthDenied invalidToken() => const AuthDenied(
        statusCode: 401,
        code: 'invalid_token',
        message: 'Debug authorization token is invalid.',
      );

  static AuthDenied tokenExpired() => const AuthDenied(
        statusCode: 401,
        code: 'token_expired',
        message: 'Debug authorization token expired.',
      );

  static AuthDenied tokenRevoked() => const AuthDenied(
        statusCode: 401,
        code: 'token_revoked',
        message: 'Debug authorization token was revoked.',
      );

  static const Set<String> _authBootstrapActions = <String>{
    'request',
    'status',
    'claim',
  };
}
